File: src/api/machines.py
from flask import Blueprint, request, jsonify, g
import logging
from sqlalchemy.exc import IntegrityError
from sqlalchemy import and_
from common.db import get_session, Machine
from common.auth import require_auth, require_role

logger = logging.getLogger(__name__)
machines_bp = Blueprint('machines', __name__)

def _is_admin() -> bool:
    return "admin" in (getattr(g, "current_roles", []) or [])

def _serialize_machine(m: Machine, admin: bool):
    # Non-admin: only minimal info (privacy and security)
    data = {
        "name": m.name,
        "reserved": m.reserved,
        "reserved_until": m.reserved_until.isoformat() if m.reserved_until else None,
    }
    if admin:
        data.update({
            "host": m.host,
            "port": m.port,
            "user": m.user,
            "reserved_by": m.reserved_by,
            "enabled": m.enabled,
            "online": m.online,
            "last_seen_at": m.last_seen_at.isoformat() if m.last_seen_at else None,
        })
    return data

@machines_bp.post("/machines")
@require_role("admin")
def add_machine():
    data = request.get_json(silent=True) or {}
    required = ["name", "host", "port", "user", "password"]
    missing = [k for k in required if not data.get(k)]
    if missing:
        return jsonify({"error": "invalid_request", "message": f"Missing fields: {', '.join(missing)}"}), 400

    session = get_session()
    try:
        m = Machine(
            name=str(data["name"]).strip(),
            host=str(data["host"]).strip(),
            port=int(data["port"]),
            user=str(data["user"]).strip(),
            password=str(data["password"]),
            reserved=False,
            reserved_by=None,
            reserved_until=None,
            enabled=True,
            online=True,
            last_seen_at=None,
        )
        session.add(m)
        session.commit()
        logger.info("machine.add name=%s host=%s port=%s", m.name, m.host, m.port)
        return jsonify(_serialize_machine(m, True)), 201
    except IntegrityError:
        session.rollback()
        return jsonify({"error": "conflict", "message": "Machine with this name already exists"}), 409
    finally:
        session.close()

@machines_bp.get("/machines")
@require_auth()
def list_machines():
    session = get_session()
    try:
        q = session.query(Machine)

        eligible = request.args.get("eligible")
        reserved = request.args.get("reserved")
        name_prefix = request.args.get("name")

        if eligible is not None:
            val = str(eligible).lower() in ("1", "true", "yes")
            if val:
                q = q.filter(and_(Machine.enabled == True, Machine.online == True, Machine.reserved == False))  # noqa: E712
            else:
                q = q.filter(~and_(Machine.enabled == True, Machine.online == True, Machine.reserved == False))  # noqa: E712
        if reserved is not None:
            val = str(reserved).lower() in ("1", "true", "yes")
            q = q.filter(Machine.reserved == val)
        if name_prefix:
            q = q.filter(Machine.name.like(f"{name_prefix}%"))

        machines = q.order_by(Machine.name.asc()).all()
        admin = _is_admin()
        out = [_serialize_machine(m, admin) for m in machines]
        logger.info("machines.list admin=%s count=%d", admin, len(out))
        return jsonify(out), 200
    finally:
        session.close()

@machines_bp.get("/machines/<name>")
@require_auth()
def get_machine(name: str):
    session = get_session()
    try:
        m = session.query(Machine).filter(Machine.name == name).one_or_none()
        if not m:
            return jsonify({"error": "not_found", "message": "Machine not found"}), 404
        admin = _is_admin()
        logger.info("machines.get name=%s admin=%s", name, admin)
        return jsonify(_serialize_machine(m, admin)), 200
    finally:
        session.close()

@machines_bp.patch("/machines/<name>")
@require_role("admin")
def update_machine(name: str):
    data = request.get_json(silent=True) or {}
    session = get_session()
    try:
        m = session.query(Machine).filter(Machine.name == name).one_or_none()
        if not m:
            return jsonify({"error": "not_found", "message": "Machine not found"}), 404

        for field in ["host", "port", "user", "password", "enabled", "online", "name"]:
            if field in data:
                setattr(m, field, data[field])
        session.commit()
        logger.info("machine.update name=%s", m.name)
        return jsonify(_serialize_machine(m, True)), 200
    finally:
        session.close()

@machines_bp.delete("/machines/<name>")
@require_role("admin")
def delete_machine(name: str):
    session = get_session()
    try:
        m = session.query(Machine).filter(Machine.name == name).one_or_none()
        if not m:
            return jsonify({"error": "not_found", "message": "Machine not found"}), 404
        if m.reserved:
            return jsonify({"error": "bad_request", "message": "Machine is currently reserved; release first"}), 400

        session.delete(m)
        session.commit()
        logger.info("machine.delete name=%s", name)
        return jsonify({"message": "deleted"}), 200
    finally:
        session.close()

File: src/api-go/internal/handlers/machines.go
package handlers

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/Jeomhps/projet-IAC/api-go/internal/db"
	"github.com/gin-gonic/gin"
)

type Machines struct{ db *db.DB }

func NewMachines(d *db.DB) *Machines { return &Machines{db: d} }

func (h *Machines) List(c *gin.Context) {
	q := "SELECT * FROM machines"
	args := []any{}
	where := []string{}

	if v := c.Query("eligible"); v != "" {
		if strings.EqualFold(v, "true") || v == "1" {
			where = append(where, "(enabled=1 AND online=1 AND reserved=0)")
		} else {
			where = append(where, "NOT (enabled=1 AND online=1 AND reserved=0)")
		}
	}
	if v := c.Query("reserved"); v != "" {
		if strings.EqualFold(v, "true") || v == "1" {
			where = append(where, "reserved=1")
		} else {
			where = append(where, "reserved=0")
		}
	}
	if v := c.Query("name"); v != "" {
		where = append(where, "name LIKE ?")
		args = append(args, v+"%")
	}
	if !c.GetBool("is_admin") {
		where = append(where, "enabled=1")
	}
	if len(where) > 0 {
		q += " WHERE " + strings.Join(where, " AND ")
	}
	q += " ORDER BY name ASC"

	var ms []db.Machine
	if err := h.db.Select(&ms, q, args...); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}

	admin := c.GetBool("is_admin")
	out := make([]gin.H, 0, len(ms))
	for _, m := range ms {
		item := gin.H{
			"name":           m.Name,
			"reserved":       m.Reserved,
			"reserved_until": formatTimePtr(m.ReservedUntil),
		}
		if admin {
			item["host"] = m.Host
			item["port"] = m.Port
			item["user"] = m.User
			item["reserved_by"] = m.ReservedBy
			item["enabled"] = m.Enabled
			item["online"] = m.Online
			item["last_seen_at"] = formatTimePtr(m.LastSeenAt)
		}
		out = append(out, item)
	}
	c.JSON(http.StatusOK, out)
}

func (h *Machines) Get(c *gin.Context) {
	name := c.Param("name")
	var m db.Machine
	if err := h.db.Get(&m, "SELECT * FROM machines WHERE name=?", name); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "not_found"})
		return
	}
	if !c.GetBool("is_admin") && !m.Enabled {
		c.JSON(http.StatusNotFound, gin.H{"error": "not_found"})
		return
	}
	admin := c.GetBool("is_admin")
	item := gin.H{
		"name": m.Name, "reserved": m.Reserved, "reserved_until": formatTimePtr(m.ReservedUntil),
	}
	if admin {
		item["host"] = m.Host
		item["port"] = m.Port
		item["user"] = m.User
		item["reserved_by"] = m.ReservedBy
		item["enabled"] = m.Enabled
		item["online"] = m.Online
		item["last_seen_at"] = formatTimePtr(m.LastSeenAt)
	}
	c.JSON(http.StatusOK, item)
}

func (h *Machines) Create(c *gin.Context) {
	var in struct {
		Name, Host, User, Password string
		Port                       int
	}
	if err := c.ShouldBindJSON(&in); err != nil || in.Name == "" || in.Host == "" || in.User == "" || in.Password == "" || in.Port <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_request"})
		return
	}
	if _, err := h.db.Exec(`INSERT INTO machines (name,host,port,user,password,reserved,reserved_by,reserved_until,enabled,online)
		VALUES (?,?,?,?,?,0,NULL,NULL,1,1)`, in.Name, in.Host, in.Port, in.User, in.Password); err != nil {
		c.JSON(http.StatusConflict, gin.H{"error": "conflict", "message": "Machine exists"})
		return
	}
	var m db.Machine
	_ = h.db.Get(&m, "SELECT * FROM machines WHERE name=?", in.Name)
	c.JSON(http.StatusCreated, gin.H{
		"name": m.Name, "host": m.Host, "port": m.Port, "user": m.User,
		"reserved": m.Reserved, "reserved_by": m.ReservedBy, "reserved_until": formatTimePtr(m.ReservedUntil),
		"enabled": m.Enabled, "online": m.Online, "last_seen_at": formatTimePtr(m.LastSeenAt),
	})
}

func (h *Machines) Update(c *gin.Context) {
	name := c.Param("name")
	var in map[string]any
	if err := c.ShouldBindJSON(&in); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_request"})
		return
	}
	fields := []string{}
	args := []any{}
	for _, f := range []string{"host", "port", "user", "password", "enabled", "online", "name"} {
		if v, ok := in[f]; ok {
			// ensure port is numeric if present
			if f == "port" {
				switch vv := v.(type) {
				case float64:
					args = append(args, int(vv))
				case string:
					if n, err := strconv.Atoi(vv); err == nil {
						args = append(args, n)
					} else {
						args = append(args, vv)
					}
				default:
					args = append(args, v)
				}
			} else {
				args = append(args, v)
			}
			fields = append(fields, f+"=?")
		}
	}
	if len(fields) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_request", "message": "no fields"})
		return
	}
	args = append(args, name)
	if _, err := h.db.Exec("UPDATE machines SET "+strings.Join(fields, ",")+" WHERE name=?", args...); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}
	h.Get(c)
}

func (h *Machines) Delete(c *gin.Context) {
	name := c.Param("name")
	var reserved bool
	if err := h.db.Get(&reserved, "SELECT reserved FROM machines WHERE name=?", name); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "not_found"})
		return
	}
	if reserved {
		c.JSON(http.StatusBadRequest, gin.H{"error": "bad_request", "message": "Machine is currently reserved; release first"})
		return
	}
	_, _ = h.db.Exec("DELETE FROM machines WHERE name=?", name)
	c.JSON(http.StatusOK, gin.H{"message": "deleted"})
}
